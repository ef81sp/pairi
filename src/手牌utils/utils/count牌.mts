import { flattenRest } from "../seek有効牌.mjs"
import type { ExtractResult5ブロック } from "../手牌utils.type.mjs"

export const count牌 = (extractResult5ブロック: ExtractResult5ブロック): number => {
  let count = 0
  if (extractResult5ブロック.雀頭) count += 2
  count += extractResult5ブロック.面子.length * 3
  count += extractResult5ブロック.塔子.length * 2
  count += flattenRest(extractResult5ブロック.rest).length
  return count
}
