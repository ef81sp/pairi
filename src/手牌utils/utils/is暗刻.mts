import type { T面子 } from "../手牌utils.type.mjs"

export const is暗刻 = (面子: T面子): boolean => {
  return (
    面子.component[0].toEqual(面子.component[1]) && 面子.component[0].toEqual(面子.component[2])
  )
}
