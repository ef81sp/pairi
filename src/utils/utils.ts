import { 牌 } from "../牌"
import { Str牌 } from "./types"

export const shuffle = (array: unknown[]) => {
  for (let i = array.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1))
    ;[array[i], array[j]] = [array[j], array[i]]
  }
}

export const generate手牌Suit別ForTest = ({
  m,
  p,
  s,
  z,
}: {
  m: (Str牌 & `${number}m`)[]
  p: (Str牌 & `${number}p`)[]
  s: (Str牌 & `${number}s`)[]
  z: (Str牌 & `${number}z`)[]
}) => ({
  m: m.map((s) => new 牌(s)),
  p: p.map((s) => new 牌(s)),
  s: s.map((s) => new 牌(s)),
  z: z.map((s) => new 牌(s)),
})
