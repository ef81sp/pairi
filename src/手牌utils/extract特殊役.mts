import { generate手牌Suit別ForTest } from "../utils/utils.mjs"
import { extract対子Tree, flatTrees } from "./extractブロックTree.mjs"
import {
  ExtractResult5ブロック,
  ExtractResult七対子,
  ExtractResult国士無双,
  T手牌Suit別,
} from "./手牌utils.type.mjs"

export const extract七対子 = (手牌: T手牌Suit別): ExtractResult七対子 => {
  const 七対子 = flatTrees(extract対子Tree(手牌))[0]
  if (七対子 === undefined) {
    return {
      対子: [],
      rest: 手牌,
    }
  }
  // ブロック内にまったく同一の対子がある場合は、片方をrestに書き戻す
  // 七対子.ブロックから重複している対子のインデックスを探す。ひとつだけでよい
  // あれば、その対子をrestに書き戻す
  const duplicatedIndex = 七対子.ブロック.findIndex((対子, index, array) => {
    return array.findIndex((対子2) => {
      return 対子.component[0].toString() === 対子2.component[0].toString()
    }) !== index
  })
  if (duplicatedIndex !== -1) {
    const duplicated対子 = 七対子.ブロック[duplicatedIndex]
    if (duplicated対子 === undefined) throw new Error("duplicated対子 is undefined")
    七対子.ブロック.splice(duplicatedIndex, 1)
    七対子.rest[duplicated対子.component[0].suit].push(duplicated対子.component[0])
    七対子.rest[duplicated対子.component[1].suit].push(duplicated対子.component[1])
  }
 

  return {
    対子: 七対子.ブロック,
    rest: 七対子.rest,
  }
}

export const extract国士無双 = (手牌: T手牌Suit別): ExtractResult国士無双 => {
  const 么九牌unique: ExtractResult国士無双["么九牌unique"] = []
  const 么九牌rest: ExtractResult国士無双["么九牌rest"] = []
  const rest: ExtractResult国士無双["rest"] = {
    m: [],
    p: [],
    s: [],
    z: [],
  }
  // m,p,sから1,9の牌を1枚ずつだけ抽出する
  for (const suit of ["m", "p", "s"] as const) {
    for (const 牌 of 手牌[suit]) {
      if (牌.number === 1 || 牌.number === 9) {
        if (!么九牌unique.find((p) => p.toString() === 牌.toString())) {
          么九牌unique.push(牌)
        } else {
          么九牌rest.push(牌)
        }
      } else {
        rest[suit].push(牌)
      }
    }
  }
  // zはすべての種類を1枚ずつだけ抽出する
  for (const 牌 of 手牌.z) {
    if (!么九牌unique.find((p) => p.toString() === 牌.toString())) {
      么九牌unique.push(牌)
    } else {
      么九牌rest.push(牌)
    }
  }
  return {
    么九牌unique,
    么九牌rest,
    rest,
  }
}
