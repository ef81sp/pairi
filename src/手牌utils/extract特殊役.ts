import { generate手牌Suit別ForTest } from "../utils/utils"
import { extract対子Tree, flatTrees } from "./extractブロックTree"
import {
  ExtractResult5ブロック,
  ExtractResult七対子,
  ExtractResult国士無双,
  T手牌Suit別,
} from "./手牌utils.type"

export const extract七対子 = (手牌: T手牌Suit別): ExtractResult七対子 => {
  const 七対子 = flatTrees(extract対子Tree(手牌))[0]
  if (七対子 === undefined) {
    return {
      対子: [],
      rest: 手牌,
    }
  }
  return {
    対子: 七対子.ブロック,
    rest: 七対子.rest,
  }
}

export const extract国士無双 = (手牌: T手牌Suit別): ExtractResult国士無双 => {
  const 么九牌unique: ExtractResult国士無双["么九牌unique"] = []
  const 么九牌rest: ExtractResult国士無双["么九牌rest"] = []
  const rest: ExtractResult国士無双["rest"] = {
    m: [],
    p: [],
    s: [],
    z: [],
  }
  // m,p,sから1,9の牌を1枚ずつだけ抽出する
  for (const suit of ["m", "p", "s"] as const) {
    for (const 牌 of 手牌[suit]) {
      if (牌.number === 1 || 牌.number === 9) {
        if (!么九牌unique.find((p) => p.toString() === 牌.toString())) {
          么九牌unique.push(牌)
        } else {
          么九牌rest.push(牌)
        }
      } else {
        rest[suit].push(牌)
      }
    }
  }
  // zはすべての種類を1枚ずつだけ抽出する
  for (const 牌 of 手牌.z) {
    if (!么九牌unique.find((p) => p.toString() === 牌.toString())) {
      么九牌unique.push(牌)
    } else {
      么九牌rest.push(牌)
    }
  }
  return {
    么九牌unique,
    么九牌rest,
    rest,
  }
}
