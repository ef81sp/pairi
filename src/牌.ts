import { Number字牌, Number数牌, Str牌, Suit, Suit字牌, Suit数牌 } from "./utils/types"

/**
 * Represents a 牌 (Mahjong tile).
 */
export class 牌 {
  readonly suit: Suit
  readonly number: typeof this.suit extends Suit字牌 ? Number字牌 : Number数牌
  readonly is赤牌: boolean
  private readonly str牌: Str牌

  /**
   * Constructs a new 牌 instance.
   * @param pai - The string representation of the 牌.
   * @param isRed - Indicates whether the 牌 is a red tile.
   */
  constructor(pai: Str牌, isRed = false) {
    const { suit, number } = parsePaiString(pai)
    this.suit = suit
    this.number = number
    this.is赤牌 = (() => {
      if (suit === "z") return false
      return number === 5 && isRed
    })()
    this.str牌 = pai
  }

  /**
   * Returns the string representation of the 牌.
   * @returns The string representation of the 牌.
   */
  toString() {
    return this.str牌
  }

  /**
   * Checks if the 牌 is equal to another 牌.
   * @param pai - The 牌 to compare.
   * @returns `true` if the 牌 is equal to the specified 牌, `false` otherwise.
   */
  toEqual(pai: 牌) {
    return this.str牌 === pai.str牌
  }

  /**
   * Creates a clone of the 牌.
   * @returns A new 牌 instance that is a clone of the current 牌.
   */
  clone() {
    return new 牌(this.str牌, this.is赤牌)
  }
}

/**
 * Parses a string representation of a 牌 and returns the corresponding suit and number.
 * @param pai - The string representation of the 牌.
 * @returns An object containing the suit and number of the 牌.
 * @throws Error if the string representation is invalid.
 */
const parsePaiString = (
  pai: string,
): { suit: Suit数牌; number: Number数牌 } | { suit: Suit字牌; number: Number字牌 } => {
  if (pai.length !== 2) throw new Error(`invalid signature ${pai}: length must be 2.`)

  const number = Number(pai[0])
  const suit = pai[1]
  const numberErrorMessage = `invalid signature ${pai}: number must be 1, 2, 3, 4, 5, 6, 7, 8 or 9.`
  const numberErrorMessage数牌 = `invalid signature ${pai}: number must be 1, 2, 3, 4, 5, 6, 7, 8 or 9.`
  const numberErrorMessage字牌 = `invalid signature ${pai}: number must be 1, 2, 3, 4, 5, 6 or 7.`

  if (!Number.isInteger(number)) throw new Error(numberErrorMessage)
  if (suit === undefined || !["m", "p", "s", "z"].includes(suit))
    throw new Error(`invalid signature ${pai}: suit must be "m", "p", "s" or "z".`)

  if (["m", "p", "s"].includes(suit)) {
    if (!(1 <= number && number <= 9)) throw new Error(numberErrorMessage数牌)
    return {
      suit,
      number,
    } as { suit: Suit数牌; number: Number数牌 }
  }
  if (!(1 <= number && number <= 7)) throw new Error(numberErrorMessage字牌)
  return {
    suit,
    number,
  } as { suit: Suit字牌; number: Number字牌 }
}

/**
 * Checks if a string is a valid 牌.
 * @param str - The string to check.
 * @returns `true` if the string is a valid 牌, `false` otherwise.
 */
export const isStr牌 = (str: string): str is Str牌 => {
  if (str.length !== 2) return false
  const number = Number(str[0])
  const suit = str[1]
  if (!Number.isInteger(number)) return false
  if (suit === undefined) return false
  if (!["m", "p", "s", "z"].includes(suit)) return false
  if (["m", "p", "s"].includes(suit)) {
    if (!(1 <= number && number <= 9)) return false
  } else {
    if (!(1 <= number && number <= 7)) return false
  }
  return true
}
