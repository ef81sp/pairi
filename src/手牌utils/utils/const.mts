import { 牌 } from "../../牌.mjs"

// biome-ignore format: みやすい
export const whole牌List: readonly 牌[] =[
  new 牌("1m"),  new 牌("2m"),  new 牌("3m"),  new 牌("4m"),  new 牌("5m"),  new 牌("6m"),  new 牌("7m"),  new 牌("8m"),  new 牌("9m"),
  new 牌("1p"),  new 牌("2p"),  new 牌("3p"),  new 牌("4p"),  new 牌("5p"),  new 牌("6p"),  new 牌("7p"),  new 牌("8p"),  new 牌("9p"),
  new 牌("1s"),  new 牌("2s"),  new 牌("3s"),  new 牌("4s"),  new 牌("5s"),  new 牌("6s"),  new 牌("7s"),  new 牌("8s"),  new 牌("9s"),
  new 牌("1z"),  new 牌("2z"),  new 牌("3z"),  new 牌("4z"),  new 牌("5z"),  new 牌("6z"),  new 牌("7z"),
]
