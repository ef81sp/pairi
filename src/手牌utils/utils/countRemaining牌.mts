import { Str牌 } from "../../utils/types.mjs"
import { 牌 } from "../../牌.mjs"

export const countRemaining牌num = (
  target: 牌[],
  visible牌List: 牌[],
): Map<Str牌, { 牌: 牌; remains: number }> => {
  const map = new Map<Str牌, { 牌: 牌; remains: number }>(
    target.map((p) => [p.toString(), { 牌: p, remains: 4 }]),
  )
  for (const visible牌 of visible牌List) {
    const p = map.get(visible牌.toString())
    if (p) {
      p.remains -= 1
    }
  }
  return map
}

if (import.meta.vitest) {
  test("全枯れ", () => {
    const result = countRemaining牌num(
      [new 牌("1m")],
      [new 牌("1m"), new 牌("1m"), new 牌("1m"), new 牌("1m")],
    )
    expect(result.get("1m")).toEqual({ 牌: new 牌("1m"), remains: 0 })
  })
}
