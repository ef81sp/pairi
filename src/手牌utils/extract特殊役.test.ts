import { generate手牌Suit別ForTest } from "../utils/utils.mjs"
import { 牌 } from "../牌.mjs"
import { extract七対子, extract国士無双 } from "./extract特殊役.mjs"

describe("七対子", () => {
  test("一向聴", () => {
    const 手牌Suit別 = generate手牌Suit別ForTest({
      m: ["1m", "1m", "4m", "4m"],
      p: ["8p", "8p", "9p", "9p"],
      s: ["5s", "6s", "7s"],
      z: ["7z", "7z"],
    })
    const result = extract七対子(手牌Suit別)
    expect(result.対子).toHaveLength(5)
    expect(result.rest).toEqual({
      m: [],
      p: [],
      s: (["5s", "6s", "7s"] as const).map((s) => new 牌(s)),
      z: [],
    })
  })

  test("聴牌", () => {
    const 手牌Suit別 = generate手牌Suit別ForTest({
      m: ["1m", "1m", "4m", "4m"],
      p: ["8p", "8p", "9p", "9p"],
      s: ["6s", "6s", "7s"],
      z: ["7z", "7z"],
    })
    const result = extract七対子(手牌Suit別)
    expect(result.対子).toHaveLength(6)
    expect(result.rest).toEqual({
      m: [],
      p: [],
      s: [new 牌("7s")],
      z: [],
    })
  })

  test("4枚使いは認められない", () => {
    const 手牌Suit別 = generate手牌Suit別ForTest({
      m: ["1m", "1m", "1m", "1m"],
      p: ["8p", "8p", "9p", "9p"],
      s: ["6s", "6s", "7s"],
      z: ["7z", "7z"],
    })
    const result = extract七対子(手牌Suit別)
    expect(result.対子).toHaveLength(5)
    expect(result.rest).toEqual({
      m: (["1m", "1m"] as const).map((s) => new 牌(s)),
      p: [],
      s: [new 牌("7s")],
      z: [],
    })
  })

  test("3枚使いは認められない", () => {
    const 手牌Suit別 = generate手牌Suit別ForTest({
      m: ["1m", "1m", "1m"],
      p: ["8p", "8p", "9p", "9p"],
      s: ["6s", "6s", "7s", "7s"],
      z: ["7z", "7z"],
    })
    const result = extract七対子(手牌Suit別)
    expect(result.対子).toHaveLength(6)
    expect(result.rest).toEqual({
      m: (["1m"] as const).map((s) => new 牌(s)),
      p: [],
      s: [],
      z: [],
    })
  })
})

test("国士無双 - 一向聴", () => {
  const 手牌Suit別 = generate手牌Suit別ForTest({
    m: ["1m", "8m", "9m"],
    p: ["1p", "9p"],
    s: ["1s", "9s"],
    z: ["1z", "2z", "3z", "4z", "5z", "6z"],
  })
  const result = extract国士無双(手牌Suit別)
  expect(result).matchSnapshot()
})
test("国士無双 - 聴牌", () => {
  const 手牌Suit別 = generate手牌Suit別ForTest({
    m: ["1m", "9m", "9m"],
    p: ["1p", "9p"],
    s: ["1s", "9s"],
    z: ["1z", "2z", "3z", "4z", "5z", "6z"],
  })
  const result = extract国士無双(手牌Suit別)
  expect(result).matchSnapshot()
})
