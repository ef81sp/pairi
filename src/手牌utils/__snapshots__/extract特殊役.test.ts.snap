// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`国士無双 - 一向聴 1`] = `
{
  "rest": {
    "m": [
      牌 {
        "is赤牌": false,
        "number": 8,
        "str牌": "8m",
        "suit": "m",
      },
    ],
    "p": [],
    "s": [],
    "z": [],
  },
  "么九牌rest": [],
  "么九牌unique": [
    牌 {
      "is赤牌": false,
      "number": 1,
      "str牌": "1m",
      "suit": "m",
    },
    牌 {
      "is赤牌": false,
      "number": 9,
      "str牌": "9m",
      "suit": "m",
    },
    牌 {
      "is赤牌": false,
      "number": 1,
      "str牌": "1p",
      "suit": "p",
    },
    牌 {
      "is赤牌": false,
      "number": 9,
      "str牌": "9p",
      "suit": "p",
    },
    牌 {
      "is赤牌": false,
      "number": 1,
      "str牌": "1s",
      "suit": "s",
    },
    牌 {
      "is赤牌": false,
      "number": 9,
      "str牌": "9s",
      "suit": "s",
    },
    牌 {
      "is赤牌": false,
      "number": 1,
      "str牌": "1z",
      "suit": "z",
    },
    牌 {
      "is赤牌": false,
      "number": 2,
      "str牌": "2z",
      "suit": "z",
    },
    牌 {
      "is赤牌": false,
      "number": 3,
      "str牌": "3z",
      "suit": "z",
    },
    牌 {
      "is赤牌": false,
      "number": 4,
      "str牌": "4z",
      "suit": "z",
    },
    牌 {
      "is赤牌": false,
      "number": 5,
      "str牌": "5z",
      "suit": "z",
    },
    牌 {
      "is赤牌": false,
      "number": 6,
      "str牌": "6z",
      "suit": "z",
    },
  ],
}
`;

exports[`国士無双 - 聴牌 1`] = `
{
  "rest": {
    "m": [],
    "p": [],
    "s": [],
    "z": [],
  },
  "么九牌rest": [
    牌 {
      "is赤牌": false,
      "number": 9,
      "str牌": "9m",
      "suit": "m",
    },
  ],
  "么九牌unique": [
    牌 {
      "is赤牌": false,
      "number": 1,
      "str牌": "1m",
      "suit": "m",
    },
    牌 {
      "is赤牌": false,
      "number": 9,
      "str牌": "9m",
      "suit": "m",
    },
    牌 {
      "is赤牌": false,
      "number": 1,
      "str牌": "1p",
      "suit": "p",
    },
    牌 {
      "is赤牌": false,
      "number": 9,
      "str牌": "9p",
      "suit": "p",
    },
    牌 {
      "is赤牌": false,
      "number": 1,
      "str牌": "1s",
      "suit": "s",
    },
    牌 {
      "is赤牌": false,
      "number": 9,
      "str牌": "9s",
      "suit": "s",
    },
    牌 {
      "is赤牌": false,
      "number": 1,
      "str牌": "1z",
      "suit": "z",
    },
    牌 {
      "is赤牌": false,
      "number": 2,
      "str牌": "2z",
      "suit": "z",
    },
    牌 {
      "is赤牌": false,
      "number": 3,
      "str牌": "3z",
      "suit": "z",
    },
    牌 {
      "is赤牌": false,
      "number": 4,
      "str牌": "4z",
      "suit": "z",
    },
    牌 {
      "is赤牌": false,
      "number": 5,
      "str牌": "5z",
      "suit": "z",
    },
    牌 {
      "is赤牌": false,
      "number": 6,
      "str牌": "6z",
      "suit": "z",
    },
  ],
}
`;
